spec/integration/aws_iam_role_spec.rb

```diff
@@ -0,0 +1,118 @@
+require 'spec_helper'
+require 'securerandom'
+
+def mk_role_name
+  name_postfix = SecureRandom.hex(8)
+  "chef_provisioning_test_iam_role_#{name_postfix}"
+end
+
+def ec2_role_policy
+<<EOF
+{
+  "Version": "2012-10-17",
+  "Statement": [
+    {
+      "Action": "sts:AssumeRole",
+      "Principal": {
+        "Service": "ec2.amazonaws.com"
+      },
+      "Effect": "Allow",
+      "Sid": ""
+    }
+  ]
+}
+EOF
+end
+
+def rds_role_policy
+<<EOF
+{
+  "Version": "2012-10-17",
+  "Statement": [
+    {
+      "Sid": "Stmt1441787971000",
+      "Effect": "Allow",
+      "Action": [
+          "rds:*"
+      ],
+      "Resource": [
+          "*"
+      ]
+    }
+  ]
+}
+EOF
+end
+
+describe Chef::Resource::AwsIamRole do
+  extend AWSSupport
+
+  when_the_chef_12_server "exists", organization: "foo", server_scope: :context do
+    with_aws "when connected to AWS" do
+
+      context "Basic IAM role creation" do
+        role_name = mk_role_name
+
+        it "aws_iam_role '#{role_name}' creates an IAM role" do
+
+          expect_recipe {
+            aws_iam_role role_name do
+              assume_role_policy_document ec2_role_policy
+            end
+          }.to create_an_aws_iam_role(role_name).and be_idempotent
+        end
+
+      end
+
+      context "create role with instance profile" do
+        role_name = mk_role_name
+
+        it "aws_iam_instance_profile '#{role_name}' creates an instance profile with role" do
+          expect_recipe {
+
+            aws_iam_role role_name do
+              assume_role_policy_document ec2_role_policy
+            end
+
+            aws_iam_instance_profile role_name do
+              role role_name
+            end
+
+          }.to create_an_aws_iam_role(role_name).and be_idempotent
+
+          role = driver.iam_resource.role(role_name)
+
+          expect(role.instance_profiles.count).to eq 1
+          expect(role.instance_profiles.first.name).to eq(role_name)
+        end
+      end
+
+      context "create role with role policy profile" do
+        role_name = mk_role_name
+
+        it "aws_iam_role_policy 'rds_full_access' creates a role policy for role" do
+          expect_recipe {
+
+            aws_iam_role role_name do
```

Comment:
It won't work. `aws_support` helper is too simple: it calls `destroy` action on resources without passing any additional attributes. `aws_iam_role_policy` requires role in order to get `aws_object`. My first attempt was to implement role policy as managed entry resource. It quickly became a huge mess and it was super hard to make it work. At that point I realised that I wrote huge pile of code _only_ to be able to automatically delete role policy inside tests using this helpers. After realising that there are no other benefits of doing so, I rolled back managed entry changes and kept as simple as it is now. In this case role polices are deleted when role is deleted.